src/Nethermind/Nethermind.Network.Discovery/Discv5/NettyDiscoveryV5Handler.cs

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Channels;
+using DotNetty.Buffers;
+using DotNetty.Transport.Channels;
+using DotNetty.Transport.Channels.Sockets;
+using Lantern.Discv5.WireProtocol.Connection;
+using Microsoft.Extensions.DependencyInjection;
+using Nethermind.Logging;
+using Nethermind.Serialization.Rlp;
+
+namespace Nethermind.Network.Discovery;
+
+/// <summary>
+/// Adapter, integrating DotNetty externally-managed <see cref="IChannel"/> with Lantern.Discv5
+/// </summary>
+public class NettyDiscoveryV5Handler : NettyDiscoveryBaseHandler, IUdpConnection
+{
+    private readonly ILogger _logger;
+    private readonly Channel<UdpReceiveResult> _inboundQueue;
+
+    private IChannel? _nettyChannel;
+
+    public NettyDiscoveryV5Handler(ILogManager loggerManager) : base(loggerManager)
+    {
+        _logger = loggerManager.GetClassLogger<NettyDiscoveryV5Handler>();
+        _inboundQueue = Channel.CreateUnbounded<UdpReceiveResult>();
+    }
+
+    public void InitializeChannel(IChannel channel) => _nettyChannel = channel;
+
+    protected override void ChannelRead0(IChannelHandlerContext ctx, DatagramPacket msg)
+    {
+        var udpPacket = new UdpReceiveResult(msg.Content.ReadAllBytesAsArray(), (IPEndPoint)msg.Sender);
+        _inboundQueue.Writer.TryWrite(udpPacket);
+    }
+
+    public async Task SendAsync(byte[] data, IPEndPoint destination)
+    {
+        if (_nettyChannel == null) throw new("Channel for discovery v5 is not initialized.");
+
+        var packet = new DatagramPacket(Unpooled.WrappedBuffer(data), destination);
+
+        try
+        {
+            await _nettyChannel.WriteAndFlushAsync(packet);
+        }
+        catch (SocketException exception)
+        {
+            if (_logger.IsError) _logger.Error($"Error sending data", exception);
+            throw;
+        }
+    }
+
+    public IAsyncEnumerable<UdpReceiveResult> ReadMessagesAsync(CancellationToken token = default) =>
+        _inboundQueue.Reader.ReadAllAsync(token);
+
+    public Task ListenAsync(CancellationToken token = default) => Task.CompletedTask;
+    public void Close() => _inboundQueue.Writer.Complete();
+
+    public static void Register(IServiceCollection services) => services
+        .AddSingleton<NettyDiscoveryV5Handler>()
```

Comment:
NP: Why not just a single `.AddSingleton<IUdpConnection, NettyDiscoveryV5Handler>`?

Reply:
`DiscoveryV5App.InitializeChannel` uses it to resolve `NettyDiscoveryV5Handler` from services directly:
```csharp
var handler = _serviceProvider.GetRequiredService<NettyDiscoveryV5Handler>();
```

Initial idea was that other internal services may be fetched from `_serviceProvider` like this, but looks like this won't be needed. Can simplify this registration and either just save `NettyDiscoveryV5Handler` instance or replace code above with:
```csharp
var handler = (NettyDiscoveryV5Handler)_serviceProvider.GetRequiredService<IUdpConnection>();
```